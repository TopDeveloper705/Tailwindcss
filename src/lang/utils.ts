export function isSpace(str: string): boolean {
  return /\s/.test(str);
}

export function isDigit(str: string): boolean {
  return /^\d+$/.test(str);
}

export function isAlpha(str: string): boolean {
  return /\w/.test(str);
}

export function connect(strings: string[]): string {
  return strings.join(';\n') + ';';
}

export function findGroupEnd(text: string, startIndex = 0): number {
  let level = 1;
  let endBracket = searchFrom(text, '}', startIndex);
  while (endBracket !== -1) {
    const nextBracket = searchFrom(text, '{', startIndex);
    if (endBracket < nextBracket || nextBracket === -1) {
      level--;
      startIndex = endBracket + 1;
      if (level === 0) return endBracket;
    } else {
      level++;
      startIndex = nextBracket + 1;
    }
    endBracket = searchFrom(text, '}', startIndex);
  }
  return -1;
}

export function searchFrom(
  text: string,
  target: string | RegExp,
  startIndex = 0,
  endIndex?: number
): number {
  // search from partial of string
  const subText = text.substring(startIndex, endIndex);
  const relativeIndex = subText.search(target);
  return relativeIndex === -1 ? -1 : startIndex + relativeIndex;
}
