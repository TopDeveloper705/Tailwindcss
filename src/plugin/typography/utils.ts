export function uniq<T>(array: T[]): T[] {
  return Array.from(new Set(array));
}

export function castArray<T>(value: T): unknown[] {
  if (!arguments.length) {
    return [];
  }
  return Array.isArray(value) ? value : [value];
}

export function isUsableColor(color: string, values: string | {[key:string]:string}): boolean {
  return Boolean(values && typeof values === 'object' && color !== 'gray' && values[600]);
}

export const round = (num: number): string =>
  num
    .toFixed(7)
    .replace(/(\.[0-9]+?)0+$/, '$1')
    .replace(/\.0$/, '');

export const rem = (px: number): string => `${round(px / 16)}rem`;

export const em = (px: number, base: number): string => `${round(px / base)}em`;
