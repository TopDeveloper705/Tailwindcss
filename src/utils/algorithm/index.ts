export { default as sortSelector } from './sortSelector';
export { default as sortMediaQuery } from './sortMediaQuery';
export { default as compileStyleSheet } from './compileStyleSheet';
export { default as cssEscape } from './cssEscape';
export { default as combineConfig } from './combineConfig';
export { default as diffConfig } from './diffConfig';
