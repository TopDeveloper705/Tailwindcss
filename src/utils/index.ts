export * from "./tools";
export * from "./helpers";
