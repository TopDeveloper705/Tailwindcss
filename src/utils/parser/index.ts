export { default as HTMLParser } from './html';
export { default as CSSParser } from './css';
export { default as ClassParser } from './class';
export { default as SelectorParser } from './selector';
