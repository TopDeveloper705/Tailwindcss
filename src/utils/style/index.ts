export { Property, InlineAtRule, Style, GlobalStyle, Keyframes } from './base';
export { StyleSheet } from './sheet';
export * from './prefixer';
