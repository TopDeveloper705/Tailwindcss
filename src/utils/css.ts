import { Property, Style, StyleSheet, InlineAtRule } from './style';
import { compile } from '../processor';
import screens from '../processor/variants/screen';

export class CSSParser {
    css:string;
    constructor(css:string) {
        this.css = css;
    }

    private _removeComment(css:string) {
        while (true) {
            const commentOpen = css.search(/\/\*/);
            const commentClose = css.search(/\*\//);
            if (commentOpen === -1 || commentClose === -1) break;
            css = css.substring(0, commentOpen) + css.substring(commentClose+2,);
        }
        return css;
    }

    private _searchGroup(text:string, startIndex=0) {
        let level = 1;
        while (true) {
            const endBracket = this._searchFrom(text, '}', startIndex);
            const nextBracket = this._searchFrom(text, '{', startIndex);
            if (endBracket === -1) return -1;
            if (endBracket < nextBracket || nextBracket === -1) {
                level --;
                startIndex = endBracket + 1;
                if (level == 0) return endBracket;
            } else {
                level ++;
                startIndex = nextBracket + 1;
            }
        }
    }

    private _searchFrom(text:string, target:string|RegExp, startIndex=0, endIndex=undefined) {
        // search from partial of string
        const subText = text.substring(startIndex, endIndex);
        const relativeIndex = subText.search(target);
        return relativeIndex === -1 ? -1 : startIndex + relativeIndex;
    }

    private _generateStyle(css:string, selector?:string, transform=false) {
        const properties:Property[] = [];
        const applies:string[] = [];
        let index = 0;
        while (true) {
            const start = this._searchFrom(css, /\S/, index);
            const end = this._searchFrom(css, ';', index);
            if (end === -1) break;
            const piece = css.substring(start, end+1);
            const isAtRule = css.charAt(start) === '@';
            const parsed = isAtRule? InlineAtRule.parse(piece) : Property.parse(piece);
            if (parsed) {
                if (transform && isAtRule && parsed.name==='apply' && parsed.value) {
                    applies.push(parsed.value);
                } else {
                    properties.push(parsed);
                }
            }
            index = end + 1;
        }
        if (applies.length > 0) {
            const styleSheet = compile(applies.join(' ')).styleSheet;
            styleSheet.children.forEach(style=>style.selector=selector);
            return (properties.length > 0) ? [new Style(selector, properties), ...styleSheet.children] : styleSheet.children;
        }
        return new Style(selector, properties);
    }

    private _clearDirectives(atrule:string):string|undefined {
        // transform @screen, keep @layer, remove other directives
        const iatrule = InlineAtRule.parse(atrule);
        if (["tailwind", "variants", "responsive"].includes(iatrule.name)) return undefined;
        if (iatrule.name === 'screen' && iatrule.value && screens.hasOwnProperty(iatrule.value)) return screens[iatrule.value]().atRules?.[0];
        return atrule;
    }

    parse(css=this.css, transform=false):StyleSheet {
        css = this._removeComment(css);
        let index = 0;
        const styleSheet = new StyleSheet();

        while (true) {
            const firstLetter = this._searchFrom(css, /\S/, index);
            if (firstLetter === -1) break;
            if (css.charAt(firstLetter) === '@') {
                // atrule
                const ruleEnd = this._searchFrom(css, ';', firstLetter);
                const nestStart = this._searchFrom(css, '{', firstLetter);

                if (nestStart === -1 || ruleEnd < nestStart) {
                    // inline atrule
                    const atrule = this._clearDirectives(css.substring(firstLetter, ruleEnd).trim());
                    if (atrule) styleSheet.add(InlineAtRule.parse(atrule).toStyle());
                    index = ruleEnd + 1;
                } else {
                    // nested atrule
                    const nestEnd = this._searchGroup(css, nestStart+1);
                    let atrule:string|undefined = css.substring(firstLetter, nestStart).trim();
                    if (transform) atrule = this._clearDirectives(atrule);
                    let result = this.parse(css.substring(nestStart + 1, nestEnd), transform).children.map(i=>i.atRule(atrule));
                    styleSheet.add(result);
                    index = nestEnd + 1;
                }
            } else {
                const nestStart = this._searchFrom(css, '{', firstLetter);
                const nestEnd = this._searchGroup(css, nestStart+1);
                if (nestStart === -1) {
                    // inline properties
                    styleSheet.add(this._generateStyle(css, undefined, transform));
                    break;
                }
                // nested selector
                styleSheet.add(this._generateStyle(css.substring(nestStart+1, nestEnd), css.substring(firstLetter, nestStart).trim(), transform));
                index = nestEnd + 1;
            }
        }
        return styleSheet;
    }
}