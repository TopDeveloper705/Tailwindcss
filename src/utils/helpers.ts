import { negateValue } from './tools';
import { Property, Style } from './style';
import type { FontSize, DarkModeConfig } from '../interfaces';

export function isString(value: unknown): value is string {
  return typeof value === 'string';
}

export function negative(scale: {
  [key: string]: string;
}): { [key: string]: string } {
  return Object.keys(scale)
    .filter((key) => scale[key] !== '0')
    .reduce(
      (negativeScale, key) => ({
        ...negativeScale,
        [`-${key}`]: negateValue(scale[key]),
      }),
      {}
    );
}

export function breakpoints(
  screens: { [key: string]: string } = {}
): { [key: string]: string } {
  return Object.keys(screens)
    .filter((key) => typeof screens[key] === 'string')
    .reduce(
      (breakpoints, key) => ({
        ...breakpoints,
        [`screen-${key}`]: screens[key],
      }),
      {}
    );
}

export function generateFontSize(font: FontSize): Property[] {
  if (typeof font === 'string') return [ new Property('font-size', font) ];
  const properties: Property[] = [];
  if (font[0]) properties.push(new Property('font-size', font[0]));
  if (typeof font[1] === 'string') {
    properties.push(new Property('line-height', font[1]));
  } else if (font[1]){
    if (font[1].lineHeight) properties.push(new Property('line-height', font[1].lineHeight));
    if (font[1].letterSpacing) properties.push(new Property('letter-spacing', font[1].letterSpacing));
  }
  return properties;
}

export function expandDirection(
  value: string,
  divide = false
): [a: string, b?: string] | undefined {
  const map: { [key: string]: [a: string, b?: string] } = {
    '': ['*'],
    y: ['top', 'bottom'],
    x: ['left', 'right'],
    t: divide ? ['top-left', 'top-right'] : ['top'],
    r: divide ? ['top-right', 'bottom-right'] : ['right'],
    b: divide ? ['bottom-right', 'bottom-left'] : ['bottom'],
    l: divide ? ['top-left', 'bottom-left'] : ['left'],
    tl: ['top-left'],
    tr: ['top-right'],
    br: ['bottom-right'],
    bl: ['bottom-left'],
  };
  if (value in map) return map[value];
}

export function generatePlaceholder(selector: string, property: Property | Property[], prefixer = false) : Style[] {
  if (!prefixer) return [ new Style(selector, property).pseudoElement('placeholder') ];
  return [
    new Style(selector, property).pseudoElement('-webkit-input-placeholder'),
    new Style(selector, property).pseudoElement('-moz-placeholder'),
    new Style(selector, property).pseudoClass('-ms-input-placeholder'),
    new Style(selector, property).pseudoElement('-ms-input-placeholder'),
    new Style(selector, property).pseudoElement('placeholder'),
  ];
}

export function toDarkStyle(style: Style, mode: DarkModeConfig): Style
export function toDarkStyle(style: Style[], mode: DarkModeConfig): Style[]
export function toDarkStyle(style: Style | Style[], mode: DarkModeConfig): Style | Style[] {
  if (!mode) return style;
  if (Array.isArray(style)) {
    if (mode === 'media') return style.map(i => new Style().atRule('@media (prefers-color-scheme: dark)').extend(i));
    return style.map(i => new Style().parent('.dark').extend(i));
  }
  if (mode === 'media') return new Style().atRule('@media (prefers-color-scheme: dark)').extend(style);
  return new Style().parent('.dark').extend(style);
}
