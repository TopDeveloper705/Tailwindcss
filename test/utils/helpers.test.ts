import {
  negative,
  breakpoints,
  generateFontSize,
  expandDirection,
} from '../../src/utils';

import { Style, StyleSheet } from '../../src/utils/style';

describe('Helpers', () => {
  it('negative', () => {
    expect(
      negative({
        px: '1px',
        1: '0.25rem',
        2: '0.5rem',
        4: '1rem',
      })
    ).toEqual({
      '-px': '-1px',
      '-1': '-0.25rem',
      '-2': '-0.5rem',
      '-4': '-1rem',
    });
  });

  it('breakpoints', () => {
    expect(
      breakpoints({
        sm: '640px',
        md: '768px',
        lg: '1024px',
        xl: '1280px',
        '2xl': '1536px',
      })
    ).toEqual({
      'screen-sm': '640px',
      'screen-md': '768px',
      'screen-lg': '1024px',
      'screen-xl': '1280px',
      'screen-2xl': '1536px',
    });
  });

  it('generateFontSize', () => {
    expect(
      new Style(
        undefined,
        generateFontSize(['0.75rem', { lineHeight: '1rem' }])
      ).build(true)
    ).toEqual('font-size:0.75rem;line-height:1rem;');
    expect(
      new Style(undefined, generateFontSize(['0.75rem'])).build(true)
    ).toEqual('font-size:0.75rem;');
    expect(
      new Style(
        undefined,
        generateFontSize([
          '0.75rem',
          { lineHeight: '1rem', letterSpacing: '2px' },
        ])
      ).build(true)
    ).toEqual('font-size:0.75rem;line-height:1rem;letter-spacing:2px;');
  });

  it('expandDirection', () => {
    expect(expandDirection('123')).toBeUndefined();
    expect(expandDirection('')).toEqual(['*']);
    expect(expandDirection('t')).toEqual(['top']);
    expect(expandDirection('t', true)).toEqual(['top-left', 'top-right']);
    expect(expandDirection('l')).toEqual(['left']);
    expect(expandDirection('l', true)).toEqual(['top-left', 'bottom-left']);
    expect(expandDirection('r')).toEqual(['right']);
    expect(expandDirection('r', true)).toEqual(['top-right', 'bottom-right']);
    expect(expandDirection('b')).toEqual(['bottom']);
    expect(expandDirection('b', true)).toEqual(['bottom-right', 'bottom-left']);
    expect(expandDirection('x')).toEqual(['left', 'right']);
    expect(expandDirection('x', true)).toEqual(['left', 'right']);
    expect(expandDirection('y')).toEqual(['top', 'bottom']);
    expect(expandDirection('y', true)).toEqual(['top', 'bottom']);
    expect(expandDirection('tl')).toEqual(['top-left']);
    expect(expandDirection('tr')).toEqual(['top-right']);
    expect(expandDirection('bl')).toEqual(['bottom-left']);
    expect(expandDirection('br')).toEqual(['bottom-right']);
  });
});
