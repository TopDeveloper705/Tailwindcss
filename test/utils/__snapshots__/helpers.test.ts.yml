Tools / generateKeyframes / bounce / 1: |-
  @keyframes bounce {
    0%, 100% {
      -webkit-transform: translateY(-25%);
      transform: translateY(-25%);
      -webkit-animation-timing-function: cubic-bezier(0.8,0,1,1);
      animation-timing-function: cubic-bezier(0.8,0,1,1);
    }
    50% {
      -webkit-transform: translateY(0);
      transform: translateY(0);
      -webkit-animation-timing-function: cubic-bezier(0,0,0.2,1);
      animation-timing-function: cubic-bezier(0,0,0.2,1);
    }
  }
  @-webkit-keyframes bounce {
    0%, 100% {
      -webkit-transform: translateY(-25%);
      transform: translateY(-25%);
      -webkit-animation-timing-function: cubic-bezier(0.8,0,1,1);
      animation-timing-function: cubic-bezier(0.8,0,1,1);
    }
    50% {
      -webkit-transform: translateY(0);
      transform: translateY(0);
      -webkit-animation-timing-function: cubic-bezier(0,0,0.2,1);
      animation-timing-function: cubic-bezier(0,0,0.2,1);
    }
  }
Tools / generateKeyframes / spin / 0: |-
  @keyframes spin {
    from {
      -webkit-transform: rotate(0deg);
      transform: rotate(0deg);
    }
    to {
      -webkit-transform: rotate(360deg);
      transform: rotate(360deg);
    }
  }
  @-webkit-keyframes spin {
    from {
      -webkit-transform: rotate(0deg);
      transform: rotate(0deg);
    }
    to {
      -webkit-transform: rotate(360deg);
      transform: rotate(360deg);
    }
  }
